### ISSUE TITLE
object-shorthand doesn't flag named functions

### ISSUE THREAD

**Issue by KaelWD** [eslint/eslint#10435]
**What rule do you want to change?**
object-shorthand

**Does this change cause the rule to produce more or fewer warnings?**
more

**How will the change be implemented? (New option, new default behavior, etc.)?**
New default behaviour for `"object-shorthand": ["error", "methods"]`

**Please provide some example code that this change will affect:**

```js
export default {
  foo: function foo () {},
  bar: function baz () {}
}
```

**What does the rule currently do for this code?**
Nothing, the rule passes

**What will the rule do after it's changed?**
`foo` should fail with "Expected method shorthand"
`bar` should pass because the function name is different, but maybe have an option to fail that too?

**Comment by g-plane** [comment]
This is working as intended. ESLint won't report named functions currently. But, should ESLint report errors if the name of function is the same as the key in object?

**Comment by platinumazure** [comment]
@g-plane We have [`func-name-matching`](https://eslint.org/docs/rules/func-name-matching) for enforcing same or different function names when the functions are on object properties.

I think I could possibly support an option on `object-shorthand` to flag cases where a method function has a name that's the same as the property and recommend shorthand. I assume we would want to continue to allow longform for when the function name is different from the property name, in general.

**Comment by g-plane** [comment]
@platinumazure You may misunderstand my opinion (sorry for my poor words), but my thought is similar to you.

**Comment by platinumazure** [comment]
Apologies @g-plane, I should have been more clear. My second paragraph wasn't directed at you in particular (it was just my general thoughts on this issue). I apologize for the miscommunication, and I also apologize for unintentionally seeming like I was assuming you had a different position.

**Comment by KaelWD** [comment]
Yeah that makes sense, I'd be fine with flagging only matching names.

**Comment by platinumazure** [comment]
@KaelWD Would you be willing to edit the original issue post to follow our [rule change proposal template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md)? That will help us evaluate a proposal to add a new option to flag matching function/property names in `object-shorthand`. Thanks!

**Comment by KaelWD** [comment]
Done.

**Comment by mysticatea** [comment]
This is working as intended. Because the shorthand method syntax doesn't have the variable of function names.

```js
let obj = {
    foo1: function foo1() {
        foo1 // Exists.
    },
    foo2() {
        foo2 // ReferenceError: foo2 is not defined.
    },
}
```

But maybe we can report it if the reference of `foo1` doesn't exist.

**Comment by KaelWD** [comment]
Huh, forgot you could do that. I guess it would have to be an option then. My use case is mainly for Vue components, where I can just call `this.foo2` if I need to access it for some reason. Our code style says shorthand only, and it would be nice if it was possible to enforce that instead of having to tell people.

**Comment by not-an-aardvark** [comment]
Thanks for your interest in improving ESLint. Unfortunately, it looks like this issue didn't get consensus from the team, so I'm closing it. We define [consensus](http://eslint.org/docs/maintainer-guide/issues.html#consensus) as having three :+1:s from team members, as well as a team member willing to champion the proposal. This is a high bar by design -- we can't realistically accept and maintain every feature request in the long term, so we only accept feature requests which are useful enough that there is consensus among the team that they're worth adding.

Since ESLint is pluggable and can load custom rules at runtime, the lack of consensus among the ESLint team doesn't need to be a blocker for you using this in your project, if you'd find it useful. It just means that you would need to [implement the rule](http://eslint.org/docs/developer-guide/working-with-rules) yourself, rather than using a bundled rule that is packaged with ESLint.
